### DISCUSSION TITLE
Tailscale Integration without JoinKey

### DISCUSSION THREAD

**Discussion by andreyrd** [k3s-io/k3s#11035] [Q&A]
**Is your feature request related to a problem? Please describe.**

I have an existing set of nodes that are already connected to Tailscale. In order to use the k3s Tailscale integration I am forced to pass through a JoinKey and have k3s manage connecting to Tailscale.

**Describe the solution you'd like**

I need k3s to manage advertising routes and such, but leave connecting to Tailscale up to me. This would mean making the joinKey optional, and if it's not present skipping the `tailscale up` command, perhaps even verifying that Tailscale is not connected.

**Describe alternatives you've considered**

I've tried to manually use the tailscale network with `--flannel-iface=tailscale0` but I'm not sure how to correctly advertise routes for pods, since I assume those can be pretty dynamic?

**Additional context**

My primary use case is that we are working towards a much more complex authentication scheme that would not be compatible with pre-authenticated keys, as well as I just don't want to be exposing authkeys in my config.

**Comment by andreyrd** [comment]
Not sure why this was converted to a question. This is a feature request. I can open a PR for it if y'all agree with my approach:

1. Remove requirement for `JoinKey` [here](https://github.com/k3s-io/k3s/blob/1ae9ca73f50d7f1b2ca4fe46ce02c1f3c307399c/pkg/vpn/vpn.go#L117)
2. And then [here](https://github.com/k3s-io/k3s/blob/1ae9ca73f50d7f1b2ca4fe46ce02c1f3c307399c/pkg/vpn/vpn.go#L51) if `JoinKey` is not specified, we can do some validation using `tailscale status` or similar to make sure it has already been manually set up by the user.

**Comment by mostpinkest** [comment]
I've done some investigating, and `tailscale up` doesn't [actually do any authentication](https://github.com/tailscale/tailscale/blob/b70c0c50fd73f134b8618792d89018cb444d8987/cmd/tailscale/cli/up.go#L569) unless `--force-reauth` is specified or the device doesn't yet have a node key. As these are existing nodes, neither requirement is met. This means you're able to set `joinKey` to any non-empty string as it won't actually used. This allows k3s to continue to manage connecting to the tailnet, while you manage authentication.

Whilst yes, requiring a joinKey to be specified is frivolous and annoying, it doesn't seem to have any effect on existing setups.

**Reply by andreyrd** [reply]
Thank you!! I'll have to give this a try soon.
